public/app/plugins/datasource/cloudwatch/components/MetricsQueryEditor.tsx

```diff
@@ -16,181 +16,132 @@ import {
 } from '../types';
 
 import QueryHeader from './QueryHeader';
+import usePreparedMetricsQuery from './usePreparedMetricsQuery';
 
 import { Alias, MathExpressionQueryField, MetricStatEditor, SQLBuilderEditor, SQLCodeEditor } from './';
 
-export type Props = QueryEditorProps<CloudWatchDatasource, CloudWatchQuery, CloudWatchJsonData>;
-
-interface State {
-  sqlCodeEditorIsDirty: boolean;
+export interface Props extends QueryEditorProps<CloudWatchDatasource, CloudWatchQuery, CloudWatchJsonData> {
+  query: CloudWatchMetricsQuery;
 }
 
-export const normalizeQuery = ({
-  namespace,
-  metricName,
-  expression,
-  dimensions,
-  region,
-  id,
-  alias,
-  statistic,
-  period,
-  sqlExpression,
-  metricQueryType,
-  metricEditorMode,
-  ...rest
-}: CloudWatchMetricsQuery): CloudWatchMetricsQuery => {
-  const normalizedQuery = {
-    queryMode: 'Metrics' as const,
-    namespace: namespace ?? '',
-    metricName: metricName ?? '',
-    expression: expression ?? '',
-    dimensions: dimensions ?? {},
-    region: region ?? 'default',
-    id: id ?? '',
-    alias: alias ?? '',
-    statistic: statistic ?? 'Average',
-    period: period ?? '',
-    metricQueryType: metricQueryType ?? MetricQueryType.Search,
-    metricEditorMode: metricEditorMode ?? MetricEditorMode.Builder,
-    sqlExpression: sqlExpression ?? '',
-    ...rest,
-  };
-  return !rest.hasOwnProperty('matchExact') ? { ...normalizedQuery, matchExact: true } : normalizedQuery;
-};
-
-export class MetricsQueryEditor extends PureComponent<Props, State> {
-  state = {
-    sqlCodeEditorIsDirty: false,
-  };
+export const MetricsQueryEditor = (props: Props) => {
+  const { query, onRunQuery, datasource } = props;
+  const [sqlCodeEditorIsDirty, setSQLCodeEditorIsDirty] = useState(false);
+  const preparedQuery = usePreparedMetricsQuery(query, props.onChange);
```

Comment:
what do you think of this suggestion:

```suggestion
useEffectOnce(() => {
  const preparedQuery = prepareQuery(query);
    if (preparedQuery !== query) {
      onChangeQuery(preparedQuery);
    }
})
```

That way you don't need to define your own hook (and I believe that the result is the same?)

Reply:
Yeah it would work. But one would need to add another hook for handling prepared query state, so feels like it comes with a bit of cluttering in the metrics editor. I actually like having a separate hook for this. It's separating concerns and makes it easier to test.